pkg/apis/extensions/validation/validation.go

```diff
@@ -346,9 +346,24 @@ func ValidateRollback(rollback *extensions.RollbackConfig, fldPath *field.Path)
 // Validates given deployment spec.
 func ValidateDeploymentSpec(spec *extensions.DeploymentSpec, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	allErrs = append(allErrs, apivalidation.ValidateNonEmptySelector(spec.Selector, fldPath.Child("selector"))...)
 	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(spec.Replicas), fldPath.Child("replicas"))...)
-	allErrs = append(allErrs, apivalidation.ValidatePodTemplateSpecForRC(&spec.Template, spec.Selector, spec.Replicas, fldPath.Child("template"))...)
+
+	if spec.Selector == nil {
+		allErrs = append(allErrs, field.Required(fldPath.Child("selector"), ""))
```

Comment:
IIUC, selector is optional only in the external version. By the time it is converted to an internal object, it is defaulted to `Spec.Template.Labels`. Since the validation happens on internal objects, validation should fail if it is nil, IMHO.

The second thing is, this code is trying to preserve the semantics of what RC based `ValidateDeploymentSpec()` did. Before, this function would call `ValidateNonEmptySelector()` and and that would call 

```
selector := labels.Set(selectorMap).AsSelector()
if selector.Empty() {
	allErrs = append(allErrs, field.Required(fldPath, ""))
}
```

And because of the way this worked, selector.Empty() would return true for nil SelectorMap and then there would be error. I am trying to preserve that behavior here.